File: fe/src/lib/services/meeting.ts
import { api } from './api';

export interface Meeting {
    id: string;
    title: string;
    channel_id: string | null;
    host_id: string | null;
    host_name: string | null;
    status: 'active' | 'ended';
    created_at: string;
    ended_at: string | null;
}

export interface MeetingParticipant {
    user_id: string;
    display_name: string;
    email: string;
    role: 'host' | 'participant';
    joined_at: string;
}

export interface MeetingAttachment {
    id: string;
    meeting_id: string;
    uploaded_by: string | null;
    uploaded_by_name: string | null;
    category: 'document' | 'recording';
    file_name: string;
    content_type: string;
    file_size: number;
    created_at: string;
}

export interface MeetingMinutes {
    meeting_id: string;
    content: string;
    updated_by: string | null;
    updated_by_name: string | null;
    updated_at: string;
}

export class MeetingApiError extends Error {
    status: number;
    constructor(message: string, status: number) {
        super(message);
        this.name = 'MeetingApiError';
        this.status = status;
    }
}

export const meetingService = {
    async createMeeting(data: { title?: string; channel_id?: string }): Promise<Meeting> {
        const res = await api.post('/meetings', data);
        if (!res.ok) throw new Error('Failed to create meeting');
        return res.json();
    },

    async getMeeting(meetingId: string): Promise<Meeting> {
        const res = await api.get(`/meetings/${meetingId}`);
        if (!res.ok) throw new Error('Failed to get meeting');
        return res.json();
    },

    async listMyMeetings(): Promise<Meeting[]> {
        const res = await api.get('/meetings');
        if (!res.ok) throw new Error('Failed to load meetings');
        return res.json();
    },

    async joinMeeting(meetingId: string): Promise<MeetingParticipant[]> {
        const res = await api.post(`/meetings/${meetingId}/join`, {});
        if (!res.ok) throw new Error('Failed to join meeting');
        return res.json();
    },

    async leaveMeeting(meetingId: string): Promise<void> {
        const res = await api.post(`/meetings/${meetingId}/leave`, {});
        if (!res.ok) throw new Error('Failed to leave meeting');
    },

    async endMeeting(meetingId: string): Promise<void> {
        const res = await api.post(`/meetings/${meetingId}/end`, {});
        if (!res.ok) throw new Error('Failed to end meeting');
    },

    async listParticipants(meetingId: string): Promise<MeetingParticipant[]> {
        const res = await api.get(`/meetings/${meetingId}/participants`);
        if (!res.ok) throw new Error('Failed to list participants');
        return res.json();
    },

    async listAttachments(meetingId: string): Promise<MeetingAttachment[]> {
        const res = await api.get(`/meetings/${meetingId}/attachments`);
        if (!res.ok) throw new MeetingApiError('Failed to load attachments', res.status);
        return res.json();
    },

    async uploadAttachment(
        meetingId: string,
        file: File,
        category?: 'document' | 'recording'
    ): Promise<MeetingAttachment> {
        const formData = new FormData();
        formData.append('file', file);
        if (category) formData.append('category', category);
        const res = await api.postForm(`/meetings/${meetingId}/attachments`, formData);
        if (!res.ok) {
            const err = await res.json().catch(() => ({}));
            throw new MeetingApiError(err.message || 'Failed to upload attachment', res.status);
        }
        return res.json();
    },

    async deleteAttachment(meetingId: string, attachmentId: string): Promise<void> {
        const res = await api.delete(`/meetings/${meetingId}/attachments/${attachmentId}`);
        if (!res.ok) throw new MeetingApiError('Failed to delete attachment', res.status);
    },

    async downloadAttachment(meetingId: string, attachmentId: string, fileName: string): Promise<void> {
        const res = await api.get(`/meetings/${meetingId}/attachments/${attachmentId}/download`);
        if (!res.ok) throw new MeetingApiError('Failed to download attachment', res.status);
        const blob = await res.blob();
        const url = URL.createObjectURL(blob);
        const link = document.createElement('a');
        link.href = url;
        link.download = fileName;
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
        URL.revokeObjectURL(url);
    },

    async getMinutes(meetingId: string): Promise<MeetingMinutes> {
        const res = await api.get(`/meetings/${meetingId}/minutes`);
        if (!res.ok) throw new MeetingApiError('Failed to load minutes', res.status);
        return res.json();
    },

    async updateMinutes(meetingId: string, content: string): Promise<MeetingMinutes> {
        const res = await api.put(`/meetings/${meetingId}/minutes`, { content });
        if (!res.ok) throw new MeetingApiError('Failed to save minutes', res.status);
        return res.json();
    }
};

File: mobile/lib/models/meeting.dart
class Meeting {
  Meeting({
    required this.id,
    required this.title,
    this.channelId,
    this.hostId,
    this.hostName,
    required this.status,
    required this.createdAt,
    this.endedAt,
  });

  final String id;
  final String title;
  final String? channelId;
  final String? hostId;
  final String? hostName;
  final String status;
  final DateTime createdAt;
  final DateTime? endedAt;

  bool get isActive => status == 'active';

  factory Meeting.fromJson(Map<String, dynamic> json) => Meeting(
        id: json['id'] as String,
        title: json['title'] as String,
        channelId: json['channel_id'] as String?,
        hostId: json['host_id'] as String?,
        hostName: json['host_name'] as String?,
        status: json['status'] as String,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
        endedAt:
            json['ended_at'] != null ? DateTime.parse(json['ended_at'] as String).toLocal() : null,
      );
}

class MeetingParticipant {
  MeetingParticipant({
    required this.userId,
    required this.displayName,
    required this.email,
    required this.role,
    required this.joinedAt,
  });

  final String userId;
  final String displayName;
  final String email;
  final String role;
  final DateTime joinedAt;

  factory MeetingParticipant.fromJson(Map<String, dynamic> json) => MeetingParticipant(
        userId: json['user_id'] as String,
        displayName: json['display_name'] as String,
        email: json['email'] as String,
        role: json['role'] as String,
        joinedAt: DateTime.parse(json['joined_at'] as String).toLocal(),
      );
}

class MeetingAttachment {
  MeetingAttachment({
    required this.id,
    required this.meetingId,
    this.uploadedById,
    this.uploadedByName,
    required this.category,
    required this.fileName,
    required this.contentType,
    required this.fileSize,
    required this.createdAt,
  });

  final String id;
  final String meetingId;
  final String? uploadedById;
  final String? uploadedByName;
  final String category;
  final String fileName;
  final String contentType;
  final int fileSize;
  final DateTime createdAt;

  factory MeetingAttachment.fromJson(Map<String, dynamic> json) => MeetingAttachment(
        id: json['id'] as String,
        meetingId: json['meeting_id'] as String,
        uploadedById: json['uploaded_by'] as String?,
        uploadedByName: json['uploaded_by_name'] as String?,
        category: json['category'] as String,
        fileName: json['file_name'] as String,
        contentType: json['content_type'] as String,
        fileSize: json['file_size'] as int,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
      );
}

class MeetingMinutes {
  MeetingMinutes({
    required this.meetingId,
    required this.content,
    this.updatedById,
    this.updatedByName,
    required this.updatedAt,
  });

  final String meetingId;
  final String content;
  final String? updatedById;
  final String? updatedByName;
  final DateTime updatedAt;

  factory MeetingMinutes.fromJson(Map<String, dynamic> json) => MeetingMinutes(
        meetingId: json['meeting_id'] as String,
        content: json['content'] as String,
        updatedById: json['updated_by'] as String?,
        updatedByName: json['updated_by_name'] as String?,
        updatedAt: DateTime.parse(json['updated_at'] as String).toLocal(),
      );
}
